### ISSUE TITLE
Expose fastify options to plugins

### ISSUE THREAD

**Issue by coreyfarrell** [fastify/fastify#1444]
## 🚀 Feature Proposal

Expose the options used to create the fastify server in a way that is accessible to plugins.

## Motivation

In some cases it might be necessary to register routes differently, for example if `ignoreTrailingSlash` is enabled it is impossible to register separate routes for `/dir` vs `/dir/`.

## Example

fastify/fastify-static#93 adds an option to send a 301 redirect when a request for a directory without the trailing slash is received.  This option currently cannot be combined with `wildcard: false` if `ignoreTrailingSlash` is enabled on the server.

```js
const fastify = require('fastify')({
  ignoreTrailingSlash: true
})

fastify.register(async (instance, opts) => {
  if (opts.initialConf.ignoreTrailingSlash) {
    // This registers `/dir` and `/dir/` so a shared handler needs to
    // determine which was requested and call handleDir or handleDirSlash.
    instance.get('/dir', handleDirOrDirSlash)
  } else {
    instance.get('/dir', handleDir)
    instance.get('/dir/', handleDirSlash)
  }
})
```

**Comment by mcollina** [comment]
Yes, I think this would be useful.

**Comment by darkgl0w** [comment]
Hi, wouldn't it be more relevant to have something like `fastify.initialConf` rather than `opt.initialConf` given that this is the initial configuration options object passed to the fastify instance that is shared between plugins.

If you want I can submit a PR to clone the options in a freezed `fastify.initalConf` object (Since `Object.freeze()` is supported since Node.js v.4.9.1, it doesn't break fastify compatibility with Node.js 6 and it will prevent users changes, making it accessible only read-only).

Are you ok with this approach ? Can I submit the PR ?

**Comment by mcollina** [comment]
I like the fastify.initialConfig approach, +1.

**Comment by jsumners** [comment]
I like this proposal as well. Particularly the freeze part.

**Comment by delvedor** [comment]
Closed in #1471.
